hs/src/Reach/EPP.hs

```diff
@@ -746,14 +778,29 @@ epp (LLProg at (LLOpts {..}) ps dli dex dvs das alias devts s) = do
   stateToSrcMap <- readIORef be_stateToSrcMap
   -- Step 4: Generate the end-points
   let SLParts {..} = ps
-  let mkep ee_who ie = do
-        let isAPI = S.member ee_who sps_apis
+  as <- readIORef be_api_steps
+  -- Ensure an API is called at most once in a given consensus step
+  forM_ (M.toAscList as) $ \ (k, v) -> do
+      forM_ (groupOn fst v) $ \ vs -> do
+        when (length vs /= 1) $ do
+          let apiAt = snd $ fromMaybe (impossible "api empty") $ headMay vs
+          expect_thrown apiAt $ Err_API_Twice k
+  -- When the same API call occurs in multiple steps,
+  -- make a seperate `EPProg` for each one.
+  let sps_ies' = M.foldrWithKey (\ k v acc ->
+          case M.lookup k as of
+            Just ns
+              | length ns > 1 -> foldr (\ (x,_) acc' -> M.insert (k, Just x) v acc') acc ns
+            _ -> M.insert (k, Nothing) v acc
+        ) mempty sps_ies
```

Comment:
I really hate `()`s in Haskell code like this and like to define a helper function beforehand